document `cfg` conditions on inline assembly templates and operands

folkertdev opened the pull request:
Reference PR for:

- https://github.com/rust-lang/rust/pull/147736

folkertdev committed:
document `cfg` conditions on inline assembly templates and operands

folkertdev commented on src/attributes.md:
```diff
@@ -108,6 +108,7 @@ Attributes may be applied to many forms in the language:
 * [Function][functions], [closure] and [function pointer]
   parameters accept outer attributes. This includes attributes on variadic parameters
   denoted with `...` in function pointers and [external blocks][variadic functions].
+* [Inline assembly] template strings and operands accept outer attributes.
```
This list is about allowed positions, and with `asm_cfg` attributes are parsed on template strings and operands, so saying that they are a valid position for attributes feels right.

But yeah they aren't really supported like in the other places listed here.

ehuss replied:
This is an interesting question because for example in things like https://github.com/rust-lang/reference/pull/1890 we document that diagnostics are allowed in all attribute positions. I'm not sure how to resolve that.

folkertdev replied:
We could move the asm line to its own section here 

```md
r[attributes.asm]
[Inline assembly] template strings and operands accept outer attributes syntactically, but 
any attribute besides `cfg` and `cfg_attr` is rejected during expansion.
```

Then asm arguments are not semantically an allowed position for attributes, they just specifically accep `cfg` and `cfg_attr`.

the section name could be `allowed-asm` too, or `allowed.asm`, I'm not sure which is best.

ehuss commented on src/inline-assembly.md:
```diff
@@ -266,6 +270,27 @@ Further constraints on the directives used by inline assembly are indicated by [
+
+r[asm.attributes.supported-attributes]
+Only the [`cfg`] and [`cfg_attr`] attributes are accepted semantically on inline assembly template strings and operands. Other attributes are parsed, but rejected when the assembly macro is expanded.
+
```
Can you add an example here showing `cfg` being used?

folkertdev replied:
I added 

```rust
core::arch::naked_asm!(
    #[cfg(not(panic = "abort"))]
    ".cfi_startproc",
    // ...
    "ret",
    #[cfg(not(panic = "abort"))]
    ".cfi_endproc",
);
```

Loosely based on https://github.com/nbdd0121/unwinding/blob/3bc28cec6c30eade0e1caa094b49db78101b19ef/src/unwinder/arch/aarch64.rs#L66C24-L66C40.

It's tough to come up with simple full examples, so hopefully this conveys the idea.

traviscross committed:
Remove stray comma from info string

There was a stray comma on this info string.  Let's remove it.

traviscross committed:
Revise `starts-with-template` example

This test is believed to fail, so we should mark it with
`compile_fail` rather than `ignore`.

Let's wrap the comment.

The compiler's error message, in this case, doesn't add any insight;
let's remove the details of that.

traviscross committed:
Remove stray comma

The sentence "Other attributes are parsed but rejected" has a two-part
compound predicate; that is, the subject is used for two verbs,
separated by a coordinating conjunction ("but"), and is not repeated.
In these cases, a comma is omitted before the coordinating
conjunction, so let's remove the comma.

traviscross committed:
Add note about restrictions on asm attributes

We have a list that describes where attributes may be applied.  We're
adding inline assembly template strings and operands to that list, but
only certain attributes may be used, and other caveats apply.  Let's
note that there are restrictions and link to the relevant section for
more details, as we do already for certain other items on this list.

It may be important, normatively, to note this restriction here if we
now or later refer to some attribute as being accepted anywhere that
attributes are allowed.

traviscross committed:
Add admonition about asm attribute implementation

The implementation of attributes for asm macro template strings and
operands doesn't use the normal system in `rustc` for handling
attributes.  This leads to the limitations and may lead to subtle
divergences in behavior.  Let's make a note about this.

For background, see:

- https://github.com/rust-lang/reference/pull/2063#discussion_r2462614284
- https://github.com/rust-lang/rust/pull/147736#issuecomment-3435781687
